### PR TITLE
Implement event on change user id

### PR DESCRIPTION
**Pull request by dtrabo**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description
[AB#59454](https://msmobilecenter.visualstudio.com/web/wi.aspx?pcguid=27bf89de-9697-418a-accf-1aa125b22914&id=59454) : [AB#59811](https://msmobilecenter.visualstudio.com/web/wi.aspx?pcguid=27bf89de-9697-418a-accf-1aa125b22914&id=59811)

Major changes:

- Added MSUserIdContextDelegate protocol
- Implemented delegate call on events of changing to a different userId
- Added corresponding coverage and unit tests to MSUserIdContextTests

## Related PRs or issues

List related PRs and other issues.

## Misc

Add what's missing, notes on what you tested, additional thoughts or questions.

### CODE REVIEW

**Comment by MatkovIvan** [comment]
@dtrabo Please **DO NOT REMOVE** template from the description. Just fill it

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.h [hunk 1]
```diff
@@ -19,11 +21,19 @@ NS_ASSUME_NONNULL_BEGIN
...
+@property(nonatomic) NSHashTable<id<MSUserIdContextDelegate>> *delegates;
+
+@property(nullable, nonatomic) NSString *mUserId;
+
 /**
  * Get singleton instance.
  */
 + (instancetype)sharedInstance;
 
+- (void)addDelegate:(id<MSUserIdContextDelegate>)delegate;
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.h]: Add comments

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.h [hunk 2]
```diff
@@ -19,11 +21,19 @@ NS_ASSUME_NONNULL_BEGIN
  */
 @property(nonatomic) NSMutableArray<MSUserIdHistoryInfo *> *userIdHistory;
 
+@property(nonatomic) NSHashTable<id<MSUserIdContextDelegate>> *delegates;
+
+@property(nullable, nonatomic) NSString *mUserId;
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.h]: Use the style from this SDK, not from Android

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m [hunk 3]
```diff
@@ -89,7 +92,22 @@ - (void)setUserId:(nullable NSString *)userId {
     [MS_USER_DEFAULTS setObject:[NSKeyedArchiver archivedDataWithRootObject:self.userIdHistory] forKey:kMSUserIdHistoryKey];
     MSLogVerbose([MSAppCenter logTag], @"Stored new userId:%@ and timestamp: %@.", self.currentUserIdInfo.userId,
                  self.currentUserIdInfo.timestamp);
+    if (![self changeUserIdIfDifferent:self.mUserId]){
+      return;
+    }
+    synchronizedDelegates = [self.delegates allObjects];
+    for (id<MSUserIdContextDelegate> delegate in synchronizedDelegates) {
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m]: Do not call delegates inside synchronized block

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m [hunk 4]
```diff
@@ -89,7 +92,22 @@ - (void)setUserId:(nullable NSString *)userId {
...
+      [delegate onNewUserId:self];
+    }
+  }
+}
+
+- (BOOL) changeUserIdIfDifferent:(NSString *)userId {
+  if (self.mUserId != nil && [self.mUserId isEqualToString:userId]) {
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m]: It is incorrect. Fix the `nil == nil` case

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContextDelegate.h [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class MSUserIdContext;
+
+@protocol MSUserIdContextDelegate <NSObject>
+
+@optional
+
+- (void)onNewUserId:(MSUserIdContext *)userId;
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContextDelegate.h]: Add comment

File: AppCenter/AppCenterTests/MSUserIdContextTests.m [hunk 6]
```diff
@@ -151,4 +150,13 @@ - (void)testPrefixedUserIdFromUserId {
   XCTAssertNil([MSUserIdContext prefixedUserIdFromUserId:nil]);
 }
 
+- (void)testUserIdEqualCurrentUserId {
```
Comment (MatkovIvan) [inline AppCenter/AppCenterTests/MSUserIdContextTests.m]: Need more tests

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.h (hunk 2 above)
Comment (Jamminroot) [reply AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.h]: I removed this one. Resolving

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m (hunk 4 above)
Comment (Jamminroot) [reply AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m]: This one is removed; Resolving

**Comment by Jamminroot** [comment]
`do not merge` is not necessary here since its a draft PR

**Comment by Jamminroot** [comment]
Note that those are only changes related to Core; further changes will be implemented in separate PR

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContextDelegate.h [hunk 7]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+NS_ASSUME_NONNULL_BEGIN
+
+@class MSUserIdContext;
+
+@protocol MSUserIdContextDelegate <NSObject>
+
+@optional
+
```
Comment (annakocheshkova) [inline AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContextDelegate.h]: Why optional? We have only one method here :) Want it - implement the protocol, don't want it - don't implement.

Comment (Jamminroot) [reply AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContextDelegate.h]: Reasonable, ill fix this. thx o/

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m [hunk 8]
```diff
@@ -61,6 +62,7 @@ - (instancetype)init {
      * Center start and setUserId call.
      */
     [MS_USER_DEFAULTS setObject:[NSKeyedArchiver archivedDataWithRootObject:self.userIdHistory] forKey:kMSUserIdHistoryKey];
+    _delegates = [NSHashTable new];
```
Comment (dhei) [inline AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m]: I think the reason we use `NSHashTable<id<MSUserIdContextDelegate>> *delegates` is to hold delegates with weak references, so you want to do it like this:

```suggestion
    _delegates = [NSHashTable weakObjectsHashTable];
```

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m [hunk 9]
```diff
@@ -83,12 +86,23 @@ - (void)setUserId:(nullable NSString *)userId {
      * crashes on apps between previous userId and current userId.
      */
     [self.userIdHistory removeLastObject];
+    BOOL sameUserId = (userId && [self.currentUserIdInfo.userId isEqualToString:(NSString *)userId]) ||
```
Comment (dhei) [inline AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m]: You can simplify like this:
```
    BOOL sameUserId = (!userId && !self.currentUserIdInfo.userId) || [self.currentUserIdInfo.userId isEqualToString:(NSString *)userId];
```

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContextDelegate.h [hunk 10]
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+NS_ASSUME_NONNULL_BEGIN
```
Comment (dhei) [inline AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContextDelegate.h]:
```suggestion
NS_ASSUME_NONNULL_BEGIN

```

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m [hunk 11]
```diff
@@ -83,12 +86,23 @@ - (void)setUserId:(nullable NSString *)userId {
...
+                      (!userId && self.currentUserIdInfo.userId == userId);
     self.currentUserIdInfo.userId = userId;
     self.currentUserIdInfo.timestamp = [NSDate date];
     [self.userIdHistory addObject:self.currentUserIdInfo];
     [MS_USER_DEFAULTS setObject:[NSKeyedArchiver archivedDataWithRootObject:self.userIdHistory] forKey:kMSUserIdHistoryKey];
     MSLogVerbose([MSAppCenter logTag], @"Stored new userId:%@ and timestamp: %@.", self.currentUserIdInfo.userId,
                  self.currentUserIdInfo.timestamp);
+    if (sameUserId) {
```
Comment (dhei) [inline AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m]: Question: if it's the same userId, wouldn't we want to return immediately instead of adding it to `self.userIdHistory` and writing to `MS_USER_DEFAULTS` ([line 93-96](https://github.com/Microsoft/AppCenter-SDK-Apple/pull/1487/files#diff-5550fe8ea22f6b4d65cedb980135be42R93)), right? 

It also seems wrong to log "Stored new userId..." message when we know it's the same UserId.

File: AppCenter/AppCenterTests/MSUserIdContextTests.m [hunk 12]
```diff
@@ -151,4 +150,50 @@ - (void)testPrefixedUserIdFromUserId {
   XCTAssertNil([MSUserIdContext prefixedUserIdFromUserId:nil]);
 }
 
+- (void)testDelegateCalledOnUserIdChanged {
```
Comment (clpolet) [inline AppCenter/AppCenterTests/MSUserIdContextTests.m]: Can we also add a test for when the UserId change to `nil`? Just extra safety.

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m (hunk 11 above)
Comment (Jamminroot) [reply AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m]: Fixed now

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m [hunk 13]
```diff
@@ -75,20 +77,29 @@ - (NSString *)userId {
 }
 
 - (void)setUserId:(nullable NSString *)userId {
+  NSArray *synchronizedDelegates;
   @synchronized(self) {
-
+    BOOL sameUserId = (!userId && !self.currentUserIdInfo.userId) || ([self.currentUserIdInfo.userId isEqualToString:(NSString *)userId]);
```
Comment (dhei) [inline AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m]:
```suggestion
    BOOL sameUserId = (!userId && !self.currentUserIdInfo.userId) || [self.currentUserIdInfo.userId isEqualToString:(NSString *)userId];
```

Remove the redundant `()`.

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m [hunk 14]
```diff
@@ -75,20 +77,29 @@ - (NSString *)userId {
...
+  NSArray *synchronizedDelegates;
   @synchronized(self) {
-
+    BOOL sameUserId = (!userId && !self.currentUserIdInfo.userId) || [self.currentUserIdInfo.userId isEqualToString:(NSString *)userId];
+    self.currentUserIdInfo.timestamp = [NSDate date];
+    if (sameUserId) {
+      return;
+    }
+    self.currentUserIdInfo.userId = userId;
     /*
```
Comment (jwallra) [inline AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m]:
```suggestion

    /*
```

File: AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m [hunk 15]
```diff
@@ -75,20 +77,30 @@ - (NSString *)userId {
 }
 
 - (void)setUserId:(nullable NSString *)userId {
+  NSArray *synchronizedDelegates;
   @synchronized(self) {
+    BOOL sameUserId = (!userId && !self.currentUserIdInfo.userId) || [self.currentUserIdInfo.userId isEqualToString:(NSString *)userId];
+    self.currentUserIdInfo.timestamp = [NSDate date];
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m]: Why do you update the timestamp here? Do you want to update this value even though it has same userId?
The `timestamp` indicates start time for the userId.

Comment (Jamminroot) [reply AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m]: exactly. We receinved same userId - so we dont need to add that to history (as pointed out by prev. comments), yet we can update info on when we did get the latest userId, which I did implement in this small refactor

Comment (jaelim-ms) [reply AppCenter/AppCenter/Internals/Context/UserId/MSUserIdContext.m]: We've synced offline for the comment. The timestamp indicates when the userId starts, it shouldn't be updated in that condition. Thanks for addressing the comment.
